File: adk/go/weil_wallet/wallet/wallet.go
// Package wallet provides secp256k1-backed wallet primitives for WeilChain.
//
// A wallet manages one or more accounts loaded from a wallet.wc file.
// Derived accounts have their secret keys re-derived from the stored xprv;
// external accounts carry their own secret keys.
//
// Use SetIndex to switch the active account at runtime.
package wallet

import (
	"bytes"
	"crypto/ecdsa"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"crypto/sha512"
	"encoding/binary"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"math/big"
	"os"

	secp "github.com/decred/dcrd/dcrec/secp256k1/v4"
	"github.com/weilliptic-public/wadk/adk/go/weil_wallet/utils"
)

// Account holds the secp256k1 key pair and the sentinel-minted account address.
type Account struct {
	SecretKey      secp.PrivateKey
	PublicKey      secp.PublicKey
	AccountAddress string
}

// GetAddress returns the sentinel-minted 72-char hex account address.
func (a *Account) GetAddress() string {
	return a.AccountAddress
}

// GetPublicKey returns the account's secp256k1 public key.
func (a *Account) GetPublicKey() secp.PublicKey {
	return a.PublicKey
}

// Sign hashes the given buffer with SHA-256 and signs with the account's
// private key, returning the hex-encoded signature.
func (a *Account) Sign(buf []byte) (*string, error) {
	digest := utils.HashSha256(buf)

	r, s, err := ecdsa.Sign(rand.Reader, a.SecretKey.ToECDSA(), digest)
	if err != nil {
		return nil, err
	}
	signature := append(r.Bytes(), s.Bytes()...)
	hexSignature := hex.EncodeToString(signature)
	return &hexSignature, nil
}

// SelectedAccount identifies which account in the wallet is currently active.
type SelectedAccount struct {
	Kind  string // "derived" or "external"
	Index int
}

// Derived returns a SelectedAccount pointing to the HD-derived account at the
// given zero-based index.
func Derived(index int) SelectedAccount {
	return SelectedAccount{Kind: "derived", Index: index}
}

// External returns a SelectedAccount pointing to the external account at the
// given zero-based index.
func External(index int) SelectedAccount {
	return SelectedAccount{Kind: "external", Index: index}
}

// OrgInfo represents an organization membership linked via `wallet link-org`.
type OrgInfo struct {
	Name     string  `json:"name"`
	Subgroup *string `json:"subgroup,omitempty"`
	Purpose  string  `json:"purpose"`
}

// orgMembershipV2 is the v2 wallet file format for org membership (no purpose).
type orgMembershipV2 struct {
	Org      string `json:"org"`
	Subgroup string `json:"subgroup"`
}

// Wallet manages derived and external accounts with a currently selected account.
type Wallet struct {
	derivedAccounts []*Account
	addedAccounts   []*Account
	currentIndex    SelectedAccount
	org             *OrgInfo
}

// ── wallet.wc file structs ────────────────────────────────────────────────────

type walletFile struct {
	Version          int                    `json:"version"`
	Type             string                 `json:"type"`
	Xprv             string                 `json:"xprv"`
	DerivedAccounts  []walletDerivedEntry   `json:"derived_accounts"`
	ExternalAccounts []walletExternalEntry  `json:"external_accounts"`
	SelectedAccount  *walletSelectedAccount `json:"selected_account"`
	Org              *OrgInfo               `json:"org,omitempty"`
	Orgs             []orgMembershipV2      `json:"orgs"`
	ActiveOrg        *int                   `json:"active_org,omitempty"`
}

type walletSelectedAccount struct {
	Type  string `json:"type"`
	Index int    `json:"index"`
}

type walletDerivedEntry struct {
	Index          int              `json:"index"`
	PublicKey      string           `json:"public_key"`
	AccountAddress string           `json:"account_address"`
	Orgs           []orgMembershipV2 `json:"orgs"`
	ActiveOrg      *int             `json:"active_org,omitempty"`
}

type walletExternalEntry struct {
	Index          int              `json:"index"`
	SecretKey      string           `json:"secret_key"`
	AccountAddress string           `json:"account_address"`
	Orgs           []orgMembershipV2 `json:"orgs"`
	ActiveOrg      *int             `json:"active_org,omitempty"`
}

// ── Constructor ───────────────────────────────────────────────────────────────

// FromWalletFile loads a Wallet from a wallet.wc file.
//
// Derived account secret keys are re-derived from the stored xprv.
// External account secret keys are read directly from the file.
// The active account is taken from the selected_account field (defaults to
// derived index 0 when absent).
func FromWalletFile(path string) (*Wallet, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("FromWalletFile: %w", err)
	}

	var wf walletFile
	if err := json.Unmarshal(data, &wf); err != nil {
		return nil, fmt.Errorf("FromWalletFile: parse: %w", err)
	}
	if wf.Type != "wallet" {
		return nil, fmt.Errorf("FromWalletFile: expected type 'wallet', got '%s'", wf.Type)
	}
	if len(wf.DerivedAccounts) == 0 && len(wf.ExternalAccounts) == 0 {
		return nil, fmt.Errorf("FromWalletFile: wallet file contains no accounts")
	}

	masterKey, masterChain, err := decodeXprv(wf.Xprv)
	if err != nil {
		return nil, fmt.Errorf("FromWalletFile: %w", err)
	}

	accountKey, accountChain, err := resolveAccountLevelKey(masterKey, masterChain, wf.DerivedAccounts)
	if err != nil {
		return nil, fmt.Errorf("FromWalletFile: %w", err)
	}

	var derivedAccounts []*Account
	for _, entry := range wf.DerivedAccounts {
		childKey, _, err := bip32DeriveChild(accountKey, accountChain, uint32(entry.Index), false)
		if err != nil {
			return nil, fmt.Errorf("FromWalletFile: derive account %d: %w", entry.Index, err)
		}
		privKey := secp.PrivKeyFromBytes(childKey[:])
		derivedAccounts = append(derivedAccounts, &Account{
			SecretKey:      *privKey,
			PublicKey:      *privKey.PubKey(),
			AccountAddress: entry.AccountAddress,
		})
	}

	var addedAccounts []*Account
	for _, entry := range wf.ExternalAccounts {
		keyBytes, err := hex.DecodeString(entry.SecretKey)
		if err != nil {
			return nil, fmt.Errorf("FromWalletFile: external account %d secret key: %w", entry.Index, err)
		}
		privKey := secp.PrivKeyFromBytes(keyBytes)
		addedAccounts = append(addedAccounts, &Account{
			SecretKey:      *privKey,
			PublicKey:      *privKey.PubKey(),
			AccountAddress: entry.AccountAddress,
		})
	}

	// Default to derived index 0 when selected_account is absent.
	kind := "derived"
	index := 0
	if wf.SelectedAccount != nil {
		kind = wf.SelectedAccount.Type
		index = wf.SelectedAccount.Index
	}

	var current SelectedAccount
	switch kind {
	case "external":
		if index >= len(addedAccounts) {
			return nil, fmt.Errorf(
				"FromWalletFile: selected external account index %d out of bounds (have %d)",
				index, len(addedAccounts),
			)
		}
		current = External(index)
	default:
		if index >= len(derivedAccounts) {
			return nil, fmt.Errorf(
				"FromWalletFile: selected derived account index %d out of bounds (have %d)",
				index, len(derivedAccounts),
			)
		}
		current = Derived(index)
	}

	// Resolve org from the selected account's per-account orgs[],
	// then fall back to top-level orgs[], then v1 org field.
	var accountOrgs []orgMembershipV2
	var accountActiveOrg *int
	switch kind {
	case "external":
		if index < len(wf.ExternalAccounts) {
			accountOrgs = wf.ExternalAccounts[index].Orgs
			accountActiveOrg = wf.ExternalAccounts[index].ActiveOrg
		}
	default:
		if index < len(wf.DerivedAccounts) {
			accountOrgs = wf.DerivedAccounts[index].Orgs
			accountActiveOrg = wf.DerivedAccounts[index].ActiveOrg
		}
	}

	var org *OrgInfo
	if len(accountOrgs) > 0 {
		// Per-account v2 orgs (preferred).
		idx := 0
		if accountActiveOrg != nil {
			idx = *accountActiveOrg
		}
		if idx < len(accountOrgs) {
			m := accountOrgs[idx]
			var subgroup *string
			if m.Subgroup != "" {
				s := m.Subgroup
				subgroup = &s
			}
			org = &OrgInfo{Name: m.Org, Subgroup: subgroup}
		}
	} else if len(wf.Orgs) > 0 {
		// Top-level v2 orgs (fallback for CLI wallet exports).
		idx := 0
		if wf.ActiveOrg != nil {
			idx = *wf.ActiveOrg
		}
		if idx < len(wf.Orgs) {
			m := wf.Orgs[idx]
			var subgroup *string
			if m.Subgroup != "" {
				s := m.Subgroup
				subgroup = &s
			}
			org = &OrgInfo{Name: m.Org, Subgroup: subgroup}
		}
	} else if wf.Org != nil {
		// v1 single org field.
		org = wf.Org
	}

	return &Wallet{
		derivedAccounts: derivedAccounts,
		addedAccounts:   addedAccounts,
		currentIndex:    current,
		org:             org,
	}, nil
}

// ── Account management ────────────────────────────────────────────────────────

// SetIndex switches the currently active account. Returns an error if the
// index is out of bounds for the specified kind.
func (w *Wallet) SetIndex(selected SelectedAccount) error {
	switch selected.Kind {
	case "derived":
		if selected.Index < 0 || selected.Index >= len(w.derivedAccounts) {
			return fmt.Errorf("derived account index %d out of range [0, %d)", selected.Index, len(w.derivedAccounts))
		}
	case "external":
		if selected.Index < 0 || selected.Index >= len(w.addedAccounts) {
			return fmt.Errorf("external account index %d out of range [0, %d)", selected.Index, len(w.addedAccounts))
		}
	default:
		return fmt.Errorf("unknown account kind: %s", selected.Kind)
	}
	w.currentIndex = selected
	return nil
}

// ── Accessors ─────────────────────────────────────────────────────────────────

// DerivedAccountCount returns the number of HD-derived accounts.
func (w *Wallet) DerivedAccountCount() int {
	return len(w.derivedAccounts)
}

// ExternalAccountCount returns the number of externally imported accounts.
func (w *Wallet) ExternalAccountCount() int {
	return len(w.addedAccounts)
}

// CurrentAccountIndex returns the currently selected account identifier.
func (w *Wallet) CurrentAccountIndex() SelectedAccount {
	return w.currentIndex
}

// GetAddress returns the sentinel-minted 72-char hex account address of the
// currently selected account.
func (w *Wallet) GetAddress() string {
	return w.currentAccount().GetAddress()
}

// Org returns the org info if this wallet has a linked organization, or nil.
func (w *Wallet) Org() *OrgInfo {
	return w.org
}

// GetPublicKey returns the secp256k1 public key of the currently selected account.
func (w *Wallet) GetPublicKey() secp.PublicKey {
	return w.currentAccount().GetPublicKey()
}

// Sign hashes the buffer and signs it with the currently selected account's key.
func (w *Wallet) Sign(buf []byte) (*string, error) {
	return w.currentAccount().Sign(buf)
}

func (w *Wallet) currentAccount() *Account {
	switch w.currentIndex.Kind {
	case "derived":
		return w.derivedAccounts[w.currentIndex.Index]
	default:
		return w.addedAccounts[w.currentIndex.Index]
	}
}

// ── BIP32 helpers ─────────────────────────────────────────────────────────────

var base58Alphabet = []byte("123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz")

// base58CheckDecode decodes a base58check string and returns the payload
// (checksum removed).
func base58CheckDecode(s string) ([]byte, error) {
	n := new(big.Int)
	for _, c := range []byte(s) {
		idx := bytes.IndexByte(base58Alphabet, c)
		if idx < 0 {
			return nil, fmt.Errorf("invalid base58 character: %c", c)
		}
		n.Mul(n, big.NewInt(58))
		n.Add(n, big.NewInt(int64(idx)))
	}

	decoded := n.Bytes()

	numLeadingZeros := 0
	for _, c := range []byte(s) {
		if c == '1' {
			numLeadingZeros++
		} else {
			break
		}
	}

	result := make([]byte, numLeadingZeros+len(decoded))
	copy(result[numLeadingZeros:], decoded)

	if len(result) < 4 {
		return nil, fmt.Errorf("base58check: payload too short")
	}
	payload := result[:len(result)-4]
	checksum := result[len(result)-4:]

	h1 := sha256.Sum256(payload)
	h2 := sha256.Sum256(h1[:])
	if !bytes.Equal(h2[:4], checksum) {
		return nil, fmt.Errorf("base58check: invalid checksum")
	}
	return payload, nil
}

// decodeXprv base58check-decodes an xprv string and returns the 32-byte
// private key and 32-byte chain code.
//
// xprv wire layout (78 bytes after checksum removal):
//
//	version(4) + depth(1) + fingerprint(4) + child_index(4)
//	+ chain_code(32) + key_prefix_0x00(1) + key(32)
func decodeXprv(xprvStr string) (key [32]byte, chain [32]byte, err error) {
	raw, err := base58CheckDecode(xprvStr)
	if err != nil {
		return key, chain, fmt.Errorf("decodeXprv: %w", err)
	}
	if len(raw) != 78 {
		return key, chain, fmt.Errorf("decodeXprv: expected 78 bytes, got %d", len(raw))
	}
	copy(chain[:], raw[13:45])
	copy(key[:], raw[46:78])
	return
}

// bip32DeriveChild derives one BIP32 child private key (hardened or normal).
func bip32DeriveChild(parentKey, parentChain [32]byte, index uint32, hardened bool) (childKey [32]byte, childChain [32]byte, err error) {
	data := make([]byte, 37)
	if hardened {
		data[0] = 0x00
		copy(data[1:33], parentKey[:])
		binary.BigEndian.PutUint32(data[33:], index+0x80000000)
	} else {
		privKey := secp.PrivKeyFromBytes(parentKey[:])
		copy(data[:33], privKey.PubKey().SerializeCompressed())
		binary.BigEndian.PutUint32(data[33:], index)
	}

	mac := hmac.New(sha512.New, parentChain[:])
	mac.Write(data)
	I := mac.Sum(nil)
	IL, IR := I[:32], I[32:]

	var ILScalar, parentScalar secp.ModNScalar
	ILScalar.SetByteSlice(IL)
	parentScalar.SetByteSlice(parentKey[:])
	ILScalar.Add(&parentScalar)

	childKeyBytes := ILScalar.Bytes()
	copy(childKey[:], childKeyBytes[:])
	copy(childChain[:], IR)
	return
}

// resolveAccountLevelKey returns the (key, chain) at the account derivation
// level. If deriving child 0 directly matches the first entry's stored
// public_key, the xprv is already at account level. Otherwise the function
// traverses m/44'/9345'/0'/0 first.
func resolveAccountLevelKey(masterKey, masterChain [32]byte, derived []walletDerivedEntry) (key [32]byte, chain [32]byte, err error) {
	if len(derived) == 0 {
		return masterKey, masterChain, nil
	}

	first := derived[0]
	childKey, _, err := bip32DeriveChild(masterKey, masterChain, uint32(first.Index), false)
	if err != nil {
		return key, chain, err
	}
	privKey := secp.PrivKeyFromBytes(childKey[:])
	pkHex := hex.EncodeToString(privKey.PubKey().SerializeCompressed())

	if pkHex == first.PublicKey {
		return masterKey, masterChain, nil
	}

	// Root xprv — traverse m/44'/9345'/0'/0.
	type step struct {
		index    uint32
		hardened bool
	}
	path := []step{{44, true}, {9345, true}, {0, true}, {0, false}}
	k, c := masterKey, masterChain
	for _, p := range path {
		k, c, err = bip32DeriveChild(k, c, p.index, p.hardened)
		if err != nil {
			return key, chain, fmt.Errorf("BIP44 path derivation: %w", err)
		}
	}
	return k, c, nil
}

File: adk/java/src/main/java/com/weilliptic/weilwallet/Wallet.java
package com.weilliptic.weilwallet;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.bitcoinj.core.ECKey;
import org.bitcoinj.crypto.ChildNumber;
import org.bitcoinj.crypto.DeterministicKey;
import org.bitcoinj.crypto.HDKeyDerivation;
import org.bitcoinj.params.MainNetParams;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;

/**
 * Multi-account secp256k1 wallet for the WeilChain platform.
 *
 * <p>Loaded from a {@code wallet.wc} file. Holds:
 * <ul>
 *   <li><b>derivedAccounts</b> — HD-derived from the {@code xprv} stored in the file.</li>
 *   <li><b>addedAccounts</b> — externally imported accounts with their own secret keys.</li>
 * </ul>
 * All signing and address operations act on the currently selected account.
 * Use {@link #setIndex(SelectedAccount)} to switch accounts at runtime.</p>
 */
public final class Wallet {

    private static final ObjectMapper JSON = new ObjectMapper();

    private final List<Account> derivedAccounts = new ArrayList<>();
    private final List<Account> addedAccounts = new ArrayList<>();
    private SelectedAccount currentIndex;
    private OrgInfo org;

    // ── Internal constructor ─────────────────────────────────────────────────

    private Wallet() {}

    /**
     * Create a Wallet from a private key and pre-minted account address.
     *
     * <p>Used by {@code derived} package and callers that hold raw keys.
     * The account is stored as {@code External(0)} and selected by default.</p>
     */
    public Wallet(PrivateKey privateKey, String accountAddress) {
        addedAccounts.add(Account.fromPrivateKeyAndAddress(privateKey, accountAddress));
        currentIndex = SelectedAccount.External(0);
    }

    // ── Static factory ───────────────────────────────────────────────────────

    /**
     * Load a Wallet from a {@code wallet.wc} file.
     *
     * <p>Derived account secret keys are re-derived from the stored {@code xprv}.
     * External account secret keys are read directly from the file.
     * The active account is taken from the {@code selected_account} field
     * (defaults to derived index 0 when absent).</p>
     *
     * @param path Path to the wallet.wc file.
     * @return A fully initialised Wallet.
     * @throws IOException              If the file cannot be read.
     * @throws IllegalArgumentException If the file type is not {@code "wallet"} or
     *                                  contains no accounts.
     */
    public static Wallet fromWalletFile(String path) throws IOException {
        return fromWalletFile(Paths.get(path));
    }

    /**
     * Load a Wallet from a {@code wallet.wc} file.
     *
     * @param path Path to the wallet.wc file.
     * @return A fully initialised Wallet.
     * @throws IOException If the file cannot be read or parsed.
     */
    public static Wallet fromWalletFile(Path path) throws IOException {
        String content = new String(Files.readAllBytes(path));
        JsonNode root = JSON.readTree(content);

        String type = root.path("type").asText("");
        if (!"wallet".equals(type)) {
            throw new IllegalArgumentException(
                "expected file type 'wallet', got '" + type + "'");
        }

        JsonNode derivedNodes   = root.path("derived_accounts");
        JsonNode externalNodes  = root.path("external_accounts");

        if ((derivedNodes.isMissingNode() || derivedNodes.size() == 0)
                && (externalNodes.isMissingNode() || externalNodes.size() == 0)) {
            throw new IllegalArgumentException("wallet file contains no accounts");
        }

        String xprvStr = root.path("xprv").asText();
        DeterministicKey masterKey = DeterministicKey.deserializeB58(xprvStr, MainNetParams.get());
        DeterministicKey accountKey = resolveAccountLevelKey(masterKey, derivedNodes);

        Wallet w = new Wallet();

        for (JsonNode entry : derivedNodes) {
            int index = entry.path("index").asInt();
            String address = entry.path("account_address").asText();
            DeterministicKey childKey = HDKeyDerivation.deriveChildKey(
                accountKey, new ChildNumber(index, false));
            ECKey ecKey = ECKey.fromPrivate(childKey.getPrivKeyBytes());
            w.derivedAccounts.add(new Account(ecKey, address));
        }

        for (JsonNode entry : externalNodes) {
            String secretKeyHex = entry.path("secret_key").asText();
            String address = entry.path("account_address").asText();
            PrivateKey pk = new PrivateKey(secretKeyHex);
            w.addedAccounts.add(Account.fromPrivateKeyAndAddress(pk, address));
        }

        // Resolve selected_account (default: derived index 0).
        String kind = "derived";
        int index = 0;
        JsonNode sel = root.path("selected_account");
        if (!sel.isMissingNode()) {
            kind  = sel.path("type").asText("derived");
            index = sel.path("index").asInt(0);
        }

        if ("external".equals(kind)) {
            if (index >= w.addedAccounts.size()) {
                throw new IllegalArgumentException(
                    "selected external account index " + index
                        + " out of bounds (have " + w.addedAccounts.size() + ")");
            }
            w.currentIndex = SelectedAccount.External(index);
        } else {
            if (index >= w.derivedAccounts.size()) {
                throw new IllegalArgumentException(
                    "selected derived account index " + index
                        + " out of bounds (have " + w.derivedAccounts.size() + ")");
            }
            w.currentIndex = SelectedAccount.Derived(index);
        }

        // ── Resolve org ─────────────────────────────────────────────────
        // Priority: per-account v2 orgs > top-level v2 orgs > v1 single org.
        w.org = resolveOrg(root, kind, index);

        return w;
    }

    // ── Account management ───────────────────────────────────────────────────

    /**
     * Switch the active account.
     *
     * @param selected The account selector (e.g. {@code SelectedAccount.Derived(1)}).
     * @throws IndexOutOfBoundsException If the index is out of bounds.
     * @throws IllegalArgumentException  If the kind is unknown.
     */
    public void setIndex(SelectedAccount selected) {
        switch (selected.kind()) {
            case "derived":
                if (selected.index() < 0 || selected.index() >= derivedAccounts.size()) {
                    throw new IndexOutOfBoundsException(
                        "derived account index " + selected.index()
                            + " out of bounds (have " + derivedAccounts.size() + " derived account(s))");
                }
                break;
            case "external":
                if (selected.index() < 0 || selected.index() >= addedAccounts.size()) {
                    throw new IndexOutOfBoundsException(
                        "external account index " + selected.index()
                            + " out of bounds (have " + addedAccounts.size() + " external account(s))");
                }
                break;
            default:
                throw new IllegalArgumentException(
                    "unknown account kind: '" + selected.kind() + "'");
        }
        currentIndex = selected;
    }

    // ── Accessors ────────────────────────────────────────────────────────────

    /** Return the number of HD-derived accounts. */
    public int derivedAccountCount() {
        return derivedAccounts.size();
    }

    /** Return the number of externally imported accounts. */
    public int externalAccountCount() {
        return addedAccounts.size();
    }

    /** Return the currently selected account index. */
    public SelectedAccount currentAccountIndex() {
        return currentIndex;
    }

    /**
     * Return the org info if this wallet has a linked organization, or {@code null} if none.
     */
    public OrgInfo org() {
        return org;
    }

    /**
     * Return the sentinel-minted 72-char hex account address of the current account.
     */
    public String getAddress() {
        return currentAccount().getAddress();
    }

    /**
     * Return the current account's secp256k1 public key (uncompressed 65 bytes for wire format).
     */
    public byte[] getPublicKeyUncompressed() {
        return currentAccount().getPublicKeyUncompressed();
    }

    /**
     * Return the underlying ECKey of the current account for low-level operations.
     */
    public ECKey getECKey() {
        return currentAccount().getSecretKey();
    }

    /**
     * Sign buf with ECDSA secp256k1 using the current account. Message is hashed
     * with SHA-256, then signed. Returns hex-encoded 64-byte compact signature (r || s).
     */
    public String sign(byte[] buf) {
        return currentAccount().sign(buf);
    }

    // ── Internal helpers ─────────────────────────────────────────────────────

    private Account currentAccount() {
        if ("derived".equals(currentIndex.kind())) {
            return derivedAccounts.get(currentIndex.index());
        }
        return addedAccounts.get(currentIndex.index());
    }

    /**
     * Resolve the active org from the wallet JSON.
     *
     * <p>Priority:
     * <ol>
     *   <li>Per-account v2 {@code orgs[]} on the selected account entry.</li>
     *   <li>Top-level v2 {@code orgs[]} with {@code active_org} index.</li>
     *   <li>v1 single {@code org} object.</li>
     * </ol>
     */
    private static OrgInfo resolveOrg(JsonNode root, String selectedKind, int selectedIndex) {
        // 1. Per-account v2 orgs on the selected account entry.
        String arrayField = "external".equals(selectedKind)
            ? "external_accounts" : "derived_accounts";
        JsonNode accountArray = root.path(arrayField);
        if (accountArray.isArray() && selectedIndex < accountArray.size()) {
            JsonNode entry = accountArray.get(selectedIndex);
            JsonNode accountOrgs = entry.path("orgs");
            if (accountOrgs.isArray() && accountOrgs.size() > 0) {
                int activeIdx = entry.path("active_org").asInt(0);
                if (activeIdx >= 0 && activeIdx < accountOrgs.size()) {
                    return orgMembershipToOrgInfo(accountOrgs.get(activeIdx));
                }
            }
        }

        // 2. Top-level v2 orgs.
        JsonNode topOrgs = root.path("orgs");
        if (topOrgs.isArray() && topOrgs.size() > 0) {
            int activeIdx = root.path("active_org").asInt(0);
            if (activeIdx >= 0 && activeIdx < topOrgs.size()) {
                return orgMembershipToOrgInfo(topOrgs.get(activeIdx));
            }
        }

        // 3. v1 single org field.
        JsonNode orgNode = root.path("org");
        if (!orgNode.isMissingNode() && orgNode.isObject()) {
            String name = orgNode.path("name").asText(null);
            if (name != null && !name.isEmpty()) {
                String subgroup = orgNode.path("subgroup").isNull()
                    ? null : orgNode.path("subgroup").asText(null);
                String purpose = orgNode.path("purpose").asText("");
                return new OrgInfo(name, subgroup, purpose);
            }
        }

        return null;
    }

    /** Convert a v2 org membership JSON node to an {@link OrgInfo}. */
    private static OrgInfo orgMembershipToOrgInfo(JsonNode node) {
        String orgName = node.path("org").asText("");
        String subgroup = node.path("subgroup").asText("");
        return new OrgInfo(
            orgName,
            subgroup.isEmpty() ? null : subgroup,
            "");
    }

    /**
     * Return the key at the account derivation level.
     *
     * <p>If deriving child 0 directly matches the first entry's stored
     * {@code public_key}, the xprv is already at account level. Otherwise
     * the method traverses {@code m/44'/9345'/0'/0} first.</p>
     */
    private static DeterministicKey resolveAccountLevelKey(
            DeterministicKey master, JsonNode derivedNodes) {

        if (!derivedNodes.isArray() || derivedNodes.size() == 0) {
            return master;
        }

        JsonNode first = derivedNodes.get(0);
        int firstIndex = first.path("index").asInt(0);
        String expectedPk = first.path("public_key").asText("");

        DeterministicKey child = HDKeyDerivation.deriveChildKey(
            master, new ChildNumber(firstIndex, false));
        byte[] compressedPub = child.getPubKeyPoint().getEncoded(true);
        String pkHex = Utils.bytesToHex(compressedPub);

        if (pkHex.equals(expectedPk)) {
            return master;
        }

        // Root xprv — traverse m/44'/9345'/0'/0.
        DeterministicKey key = master;
        key = HDKeyDerivation.deriveChildKey(key, new ChildNumber(44,   true));
        key = HDKeyDerivation.deriveChildKey(key, new ChildNumber(9345, true));
        key = HDKeyDerivation.deriveChildKey(key, new ChildNumber(0,    true));
        key = HDKeyDerivation.deriveChildKey(key, new ChildNumber(0,    false));
        return key;
    }
}
